massa-bootstrap/src/server.rs

```diff
@@ -14,363 +42,470 @@ use massa_signature::KeyPair;
     consensus_controller: Box<dyn ConsensusController>,
     network_command_sender: NetworkCommandSender,
     final_state: Arc<RwLock<FinalState>>,
-    establisher: Establisher,
-    manager_rx: mpsc::Receiver<()>,
-    bind: SocketAddr,
+    listener_rx: Receiver<BsConn>,
+    stopper_rx: Receiver<()>,
+    allow_block_list: SharedAllowBlockList<'a>,
     keypair: KeyPair,
     bootstrap_config: BootstrapConfig,
     version: Version,
     ip_hist_map: HashMap<IpAddr, Instant>,
+    runtime: Runtime,
 }
 
-#[allow(clippy::result_large_err)]
-#[allow(clippy::type_complexity)]
-fn reload_whitelist_blacklist(
-    whitelist_path: &PathBuf,
-    blacklist_path: &PathBuf,
-) -> Result<(Option<HashSet<IpAddr>>, Option<HashSet<IpAddr>>), BootstrapError> {
-    let whitelist = if let Ok(whitelist) = std::fs::read_to_string(whitelist_path) {
-        Some(
-            serde_json::from_str::<HashSet<IpAddr>>(whitelist.as_str())
-                .map_err(|_| {
-                    BootstrapError::GeneralError(String::from(
-                        "Failed to parse bootstrap whitelist",
-                    ))
-                })?
-                .into_iter()
-                .map(normalize_ip)
-                .collect(),
-        )
-    } else {
-        None
-    };
+impl BootstrapServer<'_> {
+    async fn run_updater(
+        mut list: SharedAllowBlockList<'_>,
+        interval: Duration,
+    ) -> Result<(), String> {
+        let mut interval = tokio::time::interval(interval);
+        loop {
+            interval.tick().await;
+            // TODO: loop interval here is tick + update time. Implement a state-based,
+            // rather than time-based, trigger (such as delta-count);
+            list.update()?;
+        }
+    }
 
-    let blacklist = if let Ok(blacklist) = std::fs::read_to_string(blacklist_path) {
-        Some(
-            serde_json::from_str::<HashSet<IpAddr>>(blacklist.as_str())
-                .map_err(|_| {
-                    BootstrapError::GeneralError(String::from(
-                        "Failed to parse bootstrap blacklist",
-                    ))
-                })?
-                .into_iter()
-                .map(normalize_ip)
-                .collect(),
-        )
-    } else {
-        None
-    };
-    Ok((whitelist, blacklist))
-}
+    /// Listens on a channel for incoming connections, and loads them onto a crossbeam channel
+    /// for the main-loop to process.
+    ///
+    /// Ok(Ok(())) listener closed without issue
+    /// Ok(Err((dplx, address))) listener accepted a connection then tried sending on a disconnected channel
+    /// Err(..) Error accepting a connection
+    async fn run_listener(
+        mut listener: Listener,
+        listener_tx: Sender<BsConn>,
+    ) -> Result<Result<(), BsConn>, Box<BootstrapError>> {
+        loop {
+            let msg = listener.accept().await.map_err(BootstrapError::IoError)?;
+            match listener_tx.send(msg) {
```

Comment:
Shouldn't be better to use `try_send` so that if we are full in the queue for the main loop to manage it it avoid people waiting to connect for ages ?

Reply:
In principal, I like the idea. Would have to think about the approach:

- do we immediately drop?
- do we give a short delay (maybe a second) to give the server a chance to free up a slot in the channel?
- do we make the channel unbounded so that the message is guaranteed to be processed (eventually. Would be vulnerable to clients spamming requests in such a way that flood the channel, but not the slots)
- do we give a long-ish delay, then signal the client to immediately try the next server?

...perhaps these questions will be redundant if there's a solution that involves making that channel redundant.

Reply:
I think for sending the error back to the send it should just best-effort but it's debatable